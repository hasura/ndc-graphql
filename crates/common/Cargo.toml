[package]
name = "common"
version.workspace = true
edition.workspace = true

[dependencies]
async-trait = "0.1.78"
glob-match = "0.2.1"
graphql_client = "0.14.0"
graphql-parser = "0.4.0"
ndc-models = { git = "http://github.com/hasura/ndc-spec.git", tag = "v0.1.6" }
reqwest = { version = "0.12.7", features = [
    "json",
    "rustls-tls",
], default-features = false }
schemars = "0.8.16"
serde = { version = "1.0.197", features = ["derive"] }
serde_json = "1.0.114"
tokio = "1.36.0"
