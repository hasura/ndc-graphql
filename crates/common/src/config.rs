use config_file::{RequestConfigFile, ResponseConfigFile};
use schema::SchemaDefinition;
use std::collections::BTreeMap;
pub mod config_file;
pub mod schema;

#[derive(Debug, Clone)]
pub struct ServerConfig {
    pub connection: ConnectionConfig,
    pub request: RequestConfig,
    pub response: ResponseConfig,
    pub schema: SchemaDefinition,
}

#[derive(Debug, Clone)]
pub struct ConnectionConfig {
    pub endpoint: String,
    pub headers: BTreeMap<String, String>,
}

#[derive(Debug, Clone)]
pub struct RequestConfig {
    pub headers_argument: String,
    pub headers_type_name: String,
    pub forward_headers: Vec<String>,
}
#[derive(Debug, Clone)]
pub struct ResponseConfig {
    pub headers_field: String,
    pub response_field: String,
    pub type_name_prefix: String,
    pub type_name_suffix: String,
    pub forward_headers: Vec<String>,
}

impl Default for RequestConfig {
    fn default() -> Self {
        Self {
            headers_argument: "_headers".to_owned(),
            headers_type_name: "_HeaderMap".to_owned(),
            forward_headers: vec![],
        }
    }
}

impl Default for ResponseConfig {
    fn default() -> Self {
        Self {
            headers_field: "headers".to_owned(),
            response_field: "response".to_owned(),
            type_name_prefix: "_".to_owned(),
            type_name_suffix: "Response".to_owned(),
            forward_headers: vec![],
        }
    }
}

impl From<RequestConfigFile> for RequestConfig {
    fn from(value: RequestConfigFile) -> Self {
        RequestConfig {
            headers_argument: value
                .headers_argument
                .unwrap_or_else(|| Self::default().headers_argument),
            headers_type_name: value
                .headers_type_name
                .unwrap_or_else(|| Self::default().headers_type_name),
            forward_headers: value.forward_headers.unwrap_or_default(),
        }
    }
}
impl From<ResponseConfigFile> for ResponseConfig {
    fn from(value: ResponseConfigFile) -> Self {
        ResponseConfig {
            headers_field: value
                .headers_field
                .unwrap_or_else(|| Self::default().headers_field),
            response_field: value
                .response_field
                .unwrap_or_else(|| Self::default().response_field),
            type_name_prefix: value
                .type_name_prefix
                .unwrap_or_else(|| Self::default().type_name_prefix),
            type_name_suffix: value
                .type_name_suffix
                .unwrap_or_else(|| Self::default().type_name_suffix),
            forward_headers: value.forward_headers.unwrap_or_default(),
        }
    }
}

impl ResponseConfig {
    pub fn query_response_type_name(&self, query: &str) -> String {
        format!(
            "{}{}Query{}",
            self.type_name_prefix, query, self.type_name_suffix
        )
    }
    pub fn mutation_response_type_name(&self, mutation: &str) -> String {
        format!(
            "{}{}Mutation{}",
            self.type_name_prefix, mutation, self.type_name_suffix
        )
    }
}
