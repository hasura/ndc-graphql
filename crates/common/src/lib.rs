pub mod capabilities_response;
pub mod client;
pub mod config;
pub mod schema_response;
