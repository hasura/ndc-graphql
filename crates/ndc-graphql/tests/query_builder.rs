use common::capabilities_response::capabilities_response;
use common::config::ServerConfig;
use common::{config::config_file::ServerConfigFile, schema_response::schema_response};
use insta::{assert_json_snapshot, assert_snapshot, assert_yaml_snapshot, glob};
use ndc_graphql::{
    connector::setup::GraphQLConnectorSetup,
    query_builder::{build_mutation_document, build_query_document},
};
use ndc_sdk::models;
use schemars::schema_for;
use std::{collections::HashMap, fs, path::PathBuf};

#[tokio::test]
#[ignore]
async fn update_json_schema() {
    for config in ["config-1", "config-2", "config-3"] {
        fs::write(
            format!("./tests/{config}/queries/_query_request.schema.json"),
            serde_json::to_string_pretty(&schema_for!(models::QueryRequest))
                .expect("Should serialize schema to json"),
        )
        .expect("Should be able to write out schema file");
        fs::write(
            format!("./tests/{config}/mutations/_mutation_request.schema.json"),
            serde_json::to_string_pretty(&schema_for!(models::MutationRequest))
                .expect("Should serialize schema to json"),
        )
        .expect("Should be able to write out schema file");
        fs::write(
            format!("./tests/{config}/configuration/configuration.schema.json"),
            serde_json::to_string_pretty(&schema_for!(ServerConfigFile))
                .expect("Should serialize schema to json"),
        )
        .expect("Should be able to write out schema file");
    }
}

async fn read_configuration(config: &str) -> ServerConfig {
    let configuration_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join(config)
        .join("configuration");
    let env = HashMap::from_iter(vec![
        ("GRAPHQL_ENDPOINT".to_owned(), "".to_owned()),
        ("GRAPHQL_ENDPOINT_SECRET".to_owned(), "".to_owned()),
    ]);
    GraphQLConnectorSetup::new(env)
        .read_configuration(configuration_dir)
        .await
        .expect("Should sucessfully read configuration")
}

// We use insta for snapshot testing
// Install it with `cargo install cargo-insta`
// The usual workflow is to run `cargo insta test`, then `cargo insta review`
// For more info, see insta docs: https://insta.rs/

#[tokio::test]
async fn test_build_graphql_query() {
    for config in ["config-1", "config-2", "config-3"] {
        let configuration = read_configuration(config).await;

        glob!(format!("./{config}/queries"), "*.request.json", |path| {
            let request = fs::read_to_string(path).expect("Should be able to read file");
            let request: models::QueryRequest =
                serde_json::from_str(&request).expect("Should be valid request json");
            let operation = build_query_document(&request, &configuration)
                .expect("Should sucessfully build query document");

            assert_snapshot!("Query String", operation.query);
            assert_json_snapshot!("Variables", operation.variables);
            assert_yaml_snapshot!("Headers", operation.headers);
        });
    }
}

#[tokio::test]
async fn test_build_graphql_mutation() {
    for config in ["config-1", "config-2", "config-3"] {
        let configuration = read_configuration(config).await;

        glob!(format!("./{config}/mutations"), "*.request.json", |path| {
            let request = fs::read_to_string(path).expect("Should be able to read file");
            let request: models::MutationRequest =
                serde_json::from_str(&request).expect("Should be valid request json");
            let operation = build_mutation_document(&request, &configuration)
                .expect("Should sucessfully build query document");

            assert_snapshot!("Query String", operation.query);
            assert_json_snapshot!("Variables", operation.variables);
            assert_yaml_snapshot!("Headers", operation.headers);
        });
    }
}

#[tokio::test]
async fn test_generated_schema() {
    for config in ["config-1", "config-2", "config-3"] {
        let configuration = read_configuration(config).await;
        let schema = schema_response(
            &configuration.schema,
            &configuration.request,
            &configuration.response,
        );
        assert_yaml_snapshot!(format!("{config} NDC Schema"), schema);
    }
}

#[test]
fn test_capabilities() {
    assert_yaml_snapshot!("Capabilities", capabilities_response())
}
