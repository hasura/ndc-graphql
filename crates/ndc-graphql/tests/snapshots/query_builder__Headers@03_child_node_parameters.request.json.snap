---
source: crates/ndc-graphql/tests/query_builder.rs
expression: operation.headers
input_file: crates/ndc-graphql/tests/config-3/queries/03_child_node_parameters.request.json
---
Authorization: Bearer <token>
